scipy/stats/unuran/unuran_wrapper.pyx.templ

```diff
@@ -0,0 +1,869 @@
+            size_t size = len(out)
+
+        init_unuran_callback(&callback, &unur_callback, self.callbacks,
+                             self.params)
+        if setjmp(callback.error_buf) != 0:
+            release_unuran_callback(&callback, &unur_callback)
+            return -1
+
+        for i in range(size):
+            out[i] = unur_sample_discr(rng)
+
+        release_unuran_callback(&callback, &unur_callback)
+
+        return 0
+
+    def rvs(self, size=None):
+        """
+        Sample from the distribution.
+
+        Parameters
+        ----------
+        size : int, optional
+            The shape of samples. Default is None in which case a scalar
+            sample is returned.
+
+        Returns
+        -------
+        rvs : array_like
+            A NumPy array of random variates.
+        """
+        cdef double[::1] out_cont
+        cdef int[::1] out_discr
+        N = 1 if size is None else np.prod(size)
+        if unur_distr_is_cont(unur_get_distr(self.rng)):
+            out_cont = np.empty(N, dtype=np.float64)
+            self._rvs_cont(out_cont)
+            if size is None:
+                return out_cont[0]
+            return np.asarray(out_cont).reshape(size)
+        elif unur_distr_is_discr(unur_get_distr(self.rng)):
+            out_discr = np.empty(N, dtype=np.int32)
+            self._rvs_discr(out_discr)
+            if size is None:
+                return out_discr[0]
+            return np.asarray(out_discr).reshape(size)
+        else:
+            raise NotImplementedError("only univariate continuous and "
+                                      "discrete distributions supported")
+
+    def __dealloc__(self):
+        if self.rng != NULL:
+            unur_free(self.rng)
+        if self.urng != NULL:
+            unur_urng_free(self.urng)
+
+
+cdef class TransformedDensityRejection(Method):
+    r"""
+    Transformed Density Rejection (TDR) Method.
+
+    TDR is an acceptance/rejection method that uses the concavity of a
+    transformed density to construct hat function and squeezes automatically.
+    Most universal algorithms are very slow compared to algorithms that are
+    specialized to that distribution. Algorithms that are fast have a slow
+    setup and require large tables. The aim of this universal method is to
+    provide an algorithm that is not too slow and needs only a short setup.
+    This method can be applied to continuous unimodal distributions with
+    T-concave density function. See [1]_ and [2]_ for more details.
+
+    Parameters
+    ----------
+    pdf : callable
+        PDF of the distribution. The signature of the PDF is expected to be:
+        ``def pdf(x: float, *params) -> float``. i.e. the PDF should accept
+        a Python float and the parameters and return a Python float. It need
+        not integrate to 1 i.e. the PDF doesn't need to be normalized.
+    dpdf : callable
+        Derivative of the PDF w.r.t the variate. Must have the same signature
```

Comment:
should we say "with respect to x" ?

Reply:
> should we say "with respect to x" ?

I was unsure whether writing `x` would be ambiguous. Perhaps, we can do `Derivative of the PDF w.r.t the x (variate)`. How does that sound?

Reply:
Ok!

Reply:
Done.